### ISSUE TITLE
Huge member tables cause memory issue

### ISSUE THREAD

**Issue by spekulatius** [silverstripe-australia/advancedworkflow#218]
In applications with big member tables (> 500k entries) the "Add Transition" form doesn't load up, runs out of memory and ends up with this error in the logs:

```
2015/06/27 14:46:36 [error] 1178#0: *1 FastCGI sent in stderr: "PHP message: PHP Fatal error:  Allowed memory size of 134217728 bytes exhausted (tried to allocate 74 bytes) in /var/www/.../framework/model/DataObject.php on line 388" while reading response header from upstream, client: 127.0.0.1, server: ..., request: "GET /admin/workflows/WorkflowDefinition/EditForm/field/WorkflowDefinition/item/3/ItemEditForm/field/Workflow/transition/new/62/edit HTTP/1.1", upstream: "fastcgi://127.0.0.1:9000", host: "...", referrer: ".../admin/workflows/WorkflowDefinition/EditForm/field/WorkflowDefinition/item/3/edit"
```

I followed the issue until to this point:

In the 'Form()' method in 'WorkflowFieldItemController' the 'getCMSFields()' on line 37 returns all fields based on the WorkflowTransition DataObject. Including the "Users" relation to "Member" which causes the memory issue by joining in the Member table.

Environment:
Using a nginx with PHP5-FPM and mysql.

Advancedworkflow: 3.0.4, but same issue was happening on 3.2.0
SilverStripe FW/CMS: 3.1.13

**Comment by spekulatius** [comment]
My temporary solution (removing relation and the field from the form) can be found here:

https://github.com/spekulatius/advancedworkflow/tree/temporary_fix_for_huge_member_tables

But this shouldn't be included and I'll remove the branch as soon as this issue is solved. So please fork this if you are using it somewhere.

**Comment by nyeholt** [comment]
I'm hesitant to write an advancedworkflow specific solution to this, as it's a general CMS issue in that any dropdown field bound to a dataobject type will suffer from issues if there are lots of that data type to display. The real solution is for the Dropdown / Listbox fields to support AJAX lookups for values, rather than just reading from the bound list of items.

**Comment by spekulatius** [comment]
Hello @nyeholt 

I agree, it is a general issue with DropdownFields. In this case it isn't a dropdown tho. Its a checkbox for the current user. This could be solved differently. If it would really be the problem of the dropdown I would have opened the ticket.

Spek

**Comment by nyeholt** [comment]
It is actually - CheckboxSetField extends OptionsetField which extends DropdownField :) I was actually meaning it's a broader issue that needs solving. Anyway, as for a temporary solution in your case; are you able to do the following?

- Create something like WorkflowTransitionUsersExtension, which is applied to WorkflowTransition
- define updateCMSFields($fields) in that extension class
- Create an array list of members that you _do_ want to be able to assign workflows to 
- Change the field to use that; $fields->dataFieldByName('Users')->setSource($myMembersList);

Thinking about it, a partial solution within the workflow module would be to change the $members array to just members who have cms access, though this again may be a big list.

**Comment by phptek** [comment]
@nyeholt and further narrowing it down to those users with workflow permissions ala `WorkflowService::providePermissions()`?

**Comment by spekulatius** [comment]
Sure I can do this. This would fix it much better than my solution. This solution came more from just following the problem through the module code.

> Thinking about it, a partial solution within the workflow module would be to change the $members array to just members who have cms access, though this again may be a big list. 

Sounds actually like a good way to approach this. Most users don't have access to the CMS but still in the Member table.

**Comment by icecaster** [comment]
Anyone got a decent fix for this?

**Comment by spekulatius** [comment]
Hello @icecaster ,

I have limited the users to the users which are allowed to access the workflows dashboard.

This is the branch: https://github.com/spekulatius/advancedworkflow/tree/huge-member-tables-fix-only-cms-users and this file has the change:
https://github.com/spekulatius/advancedworkflow/blob/63d456f5159f24bed983d47c5d8c871b5f861de2/code/formfields/WorkflowFieldItemController.php

@nyeholt I open a pull request for this. Let me know if you want to merge it, otherwise I'll delete it.

**Comment by phptek** [comment]
@nyeholt this looks to have been fixed in PR #223, can this be closed? @spekulatius FYI if you place "Fixes #218" into your commit log, the relevant issue is automatically dereferenced by GitHub and closed :-)

**Comment by spekulatius** [comment]
Thanks for the hint @phptek :)
